python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -163,29 +163,53 @@ def _update(self):
         # Terminate any idle or out of date nodes
         last_used = self.load_metrics.last_used_time_by_ip
         horizon = now - (60 * self.config["idle_timeout_minutes"])
+        if self.resource_demand_vector:
+            max_resources_by_ip = copy.copy(
+                self.load_metrics.get_static_node_resources_by_ip())
 
         nodes_to_terminate = []
         node_type_counts = collections.defaultdict(int)
         # Sort based on last used to make sure to keep min_workers that
         # were most recently used. Otherwise, _keep_min_workers_of_node_type
         # might keep a node that should be terminated.
         for node_id in self._sort_based_on_last_used(nodes, last_used):
-            # Make sure to not kill idle node types if the number of workers
-            # of that type is lower/equal to the min_workers of that type.
-            if self._keep_min_worker_of_node_type(
-                    node_id,
-                    node_type_counts) and self.launch_config_ok(node_id):
-                continue
-
             node_ip = self.provider.internal_ip(node_id)
-            if node_ip in last_used and last_used[node_ip] < horizon:
-                logger.info("StandardAutoscaler: "
-                            "{}: Terminating idle node".format(node_id))
-                nodes_to_terminate.append(node_id)
-            elif not self.launch_config_ok(node_id):
+            if not self.launch_config_ok(node_id):
                 logger.info("StandardAutoscaler: "
-                            "{}: Terminating outdated node".format(node_id))
+                            "{}: Terminating outdated node.".format(node_id))
                 nodes_to_terminate.append(node_id)
+                # Outdated nodes are always terminated.
+                if self.resource_demand_vector and \
+                        node_ip in max_resources_by_ip:
+                    del max_resources_by_ip[node_ip]
+            else:
+                # Do not kill idle node types if the number of workers of that
+                # type is lower/equal to the min_workers of that type.
+                if self._keep_min_worker_of_node_type(node_id,
+                                                      node_type_counts):
+                    continue
+
+                if node_ip in last_used and last_used[node_ip] < horizon:
+                    # Respecting request_resources() constraint.
+                    if self.resource_demand_vector:
+                        node_resources = copy.copy(
+                            max_resources_by_ip[node_ip])
+                        del max_resources_by_ip[node_ip]
+                        resource_requests, _ = get_bin_pack_residual(
```

Comment:
Instead of doing this in every inner step of the loop, can we instead run the bin packing once and return a list of nodes that are "needed"? This would be much faster and easier to reason about.

Reply:
I am not sure this is easier to reason about. Also, it does not increase the complexity of the scheduling as it is still within the same O().
But even if it is the case, how would you do it?
Currently, it removes a node and checks if the remainder still can fit the `request_resources()` demand.
Say you do something like this:
`unfulfilled, used_resources = binpack(non_terminated_nodes - nodes_to_terminate, requested_resources)` 
And then what? how can you know which nodes to keep? lets say it can be calculated as follow:
`_, needed_resources = binpack(nodes_to_terminate, unfulfilled)`
how do you pick the nodes to terminate from the non_terminated_nodes?